### DISCUSSION TITLE
Connections don't return in pool (SQLALchemy 2.0 + AsyncPG)

### DISCUSSION THREAD

**Discussion by thibautd** [sqlalchemy/sqlalchemy#12066] [Usage Questions]
Hello,

Technical Stack: FastAPI + SQLAchemy using an `AsyncEngine` with `asyncpg` driver.

We have a very weird bug that occurs sometimes. We are not able to reproduce it locally, but it happens in production.

Basically, sometime, after a request has been processes and the `Session` has been closed, the connection doesn't return to the pool, at least that's my guess. I've enable pool logging, and I have a mismatch between the number of logs saying "checked out from pool" and logs saying "being returned to pool".

In the end, I have the following exception:
`sqlalchemy.exc.TimeoutError: QueuePool limit of size 5 overflow 10 reached, connection timed out, timeout 30.00 (Background on this error at: https://sqlalche.me/e/20/3o7r)`

And I have to restart server to recover. I've read the documentation mentionend in the log and I don't think I'm in any of the following cases. Especially since the issue is totally random. Sometimes it doesn't happen for multiple days, sometimes it happens twice in the same day.

I cannot easily share all the code, but here are the few important parts. Please note that our transactions are always in `AUTOCOMMIT` mode by default to match Django's behavior.

</body>

</html>

**Async Engine**

```
SessionFactory: async_sessionmaker[AsyncSession] = async_sessionmaker()
Session: async_scoped_session[AsyncSession] = async_scoped_session(SessionFactory, asyncio.current_task)

def create_engine(dsn: str) -> AsyncEngine:
    engine = create_async_engine(dsn, isolation_level="AUTOCOMMIT", echo_pool=True, )
    SessionFactory.configure(bind=engine, expire_on_commit=False, close_resets_only=False)
    return engine
```

**`get_session()` FastAPI Dependency**

```
async def get_session() -> AsyncGenerator[AsyncSession, None]:
    """This dependency is used to get an AsyncSession instance and clear it.

    It doesn't start any transaction nor emit any SQL statement, it justs makes
    sure that the session is closed and removed from the sessionmaker registry.
    """
    try:
        async with Session() as session:
            yield session
    finally:
        await Session.remove()

```

**`async with transaction()` block we use to start a transaction**

```
@asynccontextmanager
async def transaction(
    isolation_level: IsolationLevel = "READ COMMITTED", clear_cache: bool = False
) -> AsyncGenerator[AsyncSession, None]:
    """Starts a transaction.

    This context manager will switch the isolation level if the connection is in AUTOCOMMIT
    mode and starts a new transaction using Session.begin() or Session.begin_nested().
    """
    session = Session()
    # session.in_transaction() is True when any statement has been issued in AUTOCOMMIT mode
    # because SQLALchemy auto-begins a virtual transaction. I didn't find a proper way to
    # check if there is a real database transaction in progress, that's why we use a custom
    # boolean to track if we are in a DB transaction or not.
    session_in_database_transaction = session.info.get("in_transaction", False)
    if session.in_transaction() and not session_in_database_transaction:
        # The session.commit() is required to update SQLAlchemy internal transaction state
        # and close the AUTOCOMMIT session, it doesn't emit any SQL statement.
        await session.commit()
    try:
        async with session.begin_nested() if session_in_database_transaction else session.begin():
            if not session_in_database_transaction:
                await session.connection(execution_options={"isolation_level": isolation_level})
                session.info["in_transaction"] = True
            if clear_cache:
                session.expunge_all()
            yield session
    finally:
        # We only do this if we created the database transaction in this block and not in a nested one.
        if not session_in_database_transaction:
            session.info["in_transaction"] = False
```

**Comment by zzzeek** [comment]
The handling of where the Session is created and closed does not seem easy to follow here.   your get_session(), that seems OK.   but then your transaction context manager, does `session = Session()` which depending on when that was called might be creating a brand new session that then goes out of the method without being closed or expunged.     not sure if that's what's happening but overall I have not been a fan of "scoped_session" for many decades for reasons like this, it's a lot safer just to pass an explicit Session around.

overall you'd need to make sure in transaction that you arent creating a brand new session.   things can also get confusing if you have requests that are raising cancelled exceptions, has been our experience particularly with fastapi.

**Reply by thibautd** [reply]
Thank you for your quick answer.

My understanding is that doing `session = Session()` always return a session bound to the current task, which _should_ be the task created by FastAPI to handle the view. But I got your point: we can't be 100% sure here, and it's error prone.

I didn't use an `async with Session()` here because I don't want to close it after the transaction() block: it will switch back to AUTOCOMMIT mode and can still be used, for example in unit tests.

I will think about removing global Session.

**Reply by zzzeek** [reply]
I would look in your logs to see if there is some precipitating event for the problem, like client connections being cancelled which would likely show some kind of stack trace in the logs.   it's the unexpected change of state from error conditions like these that often lead to this kind of problem esp with fastapi / asyncio

**Reply by thibautd** [reply]
Yeah, there is absolutely nothing in logs. Everything is working fine, no errors, nothing, and suddenly the problem is here. We'll deploy two instances and enable an agressive healthcheck to automatically restart service as a temporary workaround.

Thanks for your help, without more information I understand that it's a really hard problem to investigate.

**Reply by CaselIT** [reply]
Regarding logging maybe add tracing in `transaction` logging the id of the session (`id(session)` should be fine) to verify that it completes (the finally is run) in all cases *before* the same session is removed by `get_session`.

As a side note, in practically all cases then something like this is reported the user are using fastapi, so it seems that there is something that's not working correctly there.
